### ISSUE TITLE
BERT improvement on VBA subroutines

### ISSUE THREAD

**Issue by lampoverde** [sdllc/Basic-Excel-R-Toolkit#36]
Hello, 

I'd like to share just my thought about the interplay between Excel and R.

BERT does an outstanding job by adding R functions to Excel front-end, it's useful having these available on a spreadsheet for light functions returning single results.

**For heavy tasks** I think users would love resorting to **subroutines** as they've no dependency on formula recalculation and give better customization regarding output arrangement. I know BERT scripting capabilities would resolve, but I think it's cleaner from VBA

**Tipically such VBA subroutine**would act like an "offline" function
1. reading contents on the WB ranges
2. passing these to an R function (which might return a list in case of multiple results)
3. organizing results, by writing each list component in various parts of the WB

1) 2) are already ok, regarding 3) I'm in trouble about **resizing vectors/matrices/data.frames** (please see code below).

In general for the project, it would be nice having a **BERT utility helping "getting-out" smoothly results** from functions (the top would be preserving also results names in case of lists).

If I missed some functionality alread in BERT please excuse me. In general I think this would enhance BERT a lot.
I'd like to hear your opinion and any insight on that (a simple example follows)
Many thanks, Paolo

**R CODE**
ff<-function(v){
  list(m=mean(v),e="hello",d=v)
}

**VBA CODE**
Sub run_ff()
    a1 = Range("B3:B6")    ' Reads data (real numbers)
    x = Application.Run("BERT.Call", "ff", a1)     'run ff, x is variant loosing list names (Top: preseving names)
   ' Writes results
    Range("A1") = x(1, 1)               ' ok it works, scalar number
    Range("A2") = x(2, 1)               ' ok it works, scalar string
   'Range("A3:A6") = x(3, 1)          'this would work (but one needs to specify range size)
   Call getRes(x(3, 1), Range("A3"))   ' Top: having a BERT help-function reading arg and resizing destination
End Sub

# #

**Comment by duncanwerner** [comment]
One thing to understand is that BERT functions are non-volatile.  That means they only recalculate when you change the inputs.  For some things VBA is useful but for the most part it's not necessary.

We can't preserve names in lists (there actually might be a way, but it's not practical because it won't work in the spreadsheet).  As far as resizing the result, why not just do it the way Alex did it in the selection add-in?  That seems to work fine, it's just a few lines of VBA.

**Comment by lampoverde** [comment]
I understand your point, my view could be biased as I use R and Excel for
heavy tasks.
At the moment I cannot find a way to mimic a VBA sub with a function as you
pointed out. I will try to understand if it's possible.

Suppose I put into a cell one R.function which depends on 10 inputs and
takes 5' to complete one run and returns 5 results.

If I need to take one new run by changing some inputs, every time I change
one of them the function starts running (switching to manual formula in
excel options would solve, but I don't find it natural). Also each time I
open the WB the function would start running.

The last point is about multiple results, if I need to spread function
results over an organized spreadsheet like a dashboard, through VBA it's
pretty natural.

Thanks for your feedback on R list names on variant, given that now I will
follow Alex VBA.




2017-03-07 18:20 GMT+01:00 duncanwerner <notifications@github.com>:

**Comment by duncanwerner** [comment]
VBA is certainly good for some cases.  I think you're correct in your case that it makes more sense.

However I'm worried about a task that takes 5 minutes -- are you really running things for 5 minutes, or was that just an example?  If you have a single task that runs for 5 minutes, it might have problems in the current structure.  The reason is that windows applications need to deal with messages/events every once in a while.  BERT runs R in process, so a 5 minute R function will block this message handling.

If your function is iterative, one way to handle it is to iterate from VBA.  That's wasteful, though, as it requires a lot of overhead for each iteration.  It would be better if we had some framework for long tasks that could provide progress reporting (and perhaps support canceling).

**Comment by lampoverde** [comment]
Yes it may take 3 to 5 mins, it's about Monte Carlo modelling. You touched
the key-point about running these functions (they freeze R for a long time
while one would need at least a mechanism to exit loops gracefully).

So if I understood, you're stating that if R "blocks" Excel/VBA for a long
time problems may arise with Excel? My current experience is with
Excel/Matlab, I don't remember any crash, for sure it's annoying as it
freezes Excel and one need to wait a long time without the possibility to
use it.

Regarding the possibility to interrupt loops one possibility is to use RCPP
+ RcppProgress.

Regarding achieve asynchronous computing I'd like to investigate the
"future" package but it's too early, it's just an idea I don't know if it
works properly.

I was impressed about the way F# allows asynchronous functions in Excel,
for anyone interested here's a video about FCell
https://youtu.be/iWvIpDnDIug?t=322


Many thanks for the discussion













2017-03-08 17:28 GMT+01:00 duncanwerner <notifications@github.com>:
